### PR TITLE
2113: In stock, 0 qty and backorders allowed products not visible on category pages

### PR DESCRIPTION
**Pull request by malesh-v**
### Fixed Issue
magento-engcom/msi#2113: In stock, 0 qty and backorders allowed products not visible on category pages

### Manual testing scenarios (*)
1. Create custom stock with custom source and main website assigned
2. Create a simple product with custom source assigned with qty '0' and status 'In Stock'
3. Go to Advanced Inventory
4. Set Manage Stock to Yes
5. Set Backorders to 'Allow Qty Below 0 And Notify Customer'

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento-engcom/msi?pullRequest=2115) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dev.magento.com/magento-engcom/msi?pullRequest=2115) before we can accept your contribution.<br/><hr/>**vadim** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).

File: app/code/Magento/InventorySales/Model/ResourceModel/IsStockItemSalableCondition/BackordersCondition.php [hunk 1]
```diff
@@ -35,11 +35,14 @@ public function __construct(StockConfigurationInterface $configuration)
...
+        $useDefaultBackorders = 'legacy_stock_item.use_config_backorders';
+        $itemMinQty = 'legacy_stock_item.min_qty';
 
-        $condition = (1 === $globalBackorders)
-            ? 'legacy_stock_item.use_config_backorders = 1'
-            : 'legacy_stock_item.use_config_backorders = 0 AND legacy_stock_item.backorders = 1';
-        $condition .= ' AND (legacy_stock_item.min_qty >= 0 OR legacy_stock_item.qty > legacy_stock_item.min_qty)';
+        $condition = $globalBackorders === 0
+            ? "$useDefaultBackorders = 0 AND $itemBackordersCondition"
+            : "$useDefaultBackorders = 1 OR $itemBackordersCondition";
```
Comment (ishakhsuvarov) [inline app/code/Magento/InventorySales/Model/ResourceModel/IsStockItemSalableCondition/BackordersCondition.php:44]: How is additional `OR` clause used here? I believe the only problem was that `backorders` may be not only `0` and `1` but `2` as well, it seems to be covered by updating the `AND` clause above

**Review by ishakhsuvarov — CHANGES_REQUESTED** [review]
* Please add or adjust tests to support this case.
* Please sign CLA

File: app/code/Magento/InventorySales/Model/ResourceModel/IsStockItemSalableCondition/BackordersCondition.php (hunk 1 above)
Comment (malesh-v) [reply app/code/Magento/InventorySales/Model/ResourceModel/IsStockItemSalableCondition/BackordersCondition.php:44]: `OR `covers next case : 
* in M2 admin general settings _backorders_ is 1 or 2;
* on product _backorders_ is also 1 or 2 and 'Use default' for _backorders_ is **unchecked**;
* finally product won't present.

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento-engcom/msi?pullRequest=2115) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.dev.magento.com/magento-engcom/msi?pullRequest=2115) before we can accept your contribution.<br/>**2** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: ishakhsuvarov<br/>:white_check_mark: vad1m777<br/>:x: vadim<hr/>**vadim** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).

File: app/code/Magento/InventoryApi/Test/_files/stocks_with_sales_channel.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
+<?php
```
Comment (ishakhsuvarov) [inline app/code/Magento/InventoryApi/Test/_files/stocks_with_sales_channel.php:1]: Fixture `app/code/Magento/InventorySalesApi/Test/_files/stock_website_sales_channels.php` seems to be similar. Please investigate if it's possible to use it.

File: app/code/Magento/InventorySales/Test/Integration/IsProductSalable/BackorderConditionTest.php [hunk 3]
```diff
@@ -18,9 +18,32 @@
...
 use PHPUnit\Framework\TestCase;
+use Magento\Catalog\Model\ResourceModel\Product\Collection;
+use Magento\Framework\App\Config\Storage\WriterInterface;
+use Magento\InventoryIndexer\Indexer\InventoryIndexer;
+use Magento\Framework\App\Config;
 
 class BackorderConditionTest extends TestCase
 {
+    private const PRODUCT_SKU = 'SKU-2';
+    private const NEGATIVE_QTY = -15;
```
Comment (ishakhsuvarov) [inline app/code/Magento/InventorySales/Test/Integration/IsProductSalable/BackorderConditionTest.php:29]: Usually `QTY` is a float

File: app/code/Magento/InventorySales/Test/Integration/IsProductSalable/BackorderConditionTest.php [hunk 4]
```diff
@@ -67,40 +95,53 @@ protected function setUp()
...
+     *
+     * @return void
      */
-    public function testBackorderedZeroQtyProductIsSalable()
-    {
-        $product = $this->productRepository->get('SKU-2');
+    public function testBackorderedZeroQtyProductIsSalable(
+        int $backorders,
+        int $useDefaultBackOrders,
+        int $exResult
```
Comment (ishakhsuvarov) [inline app/code/Magento/InventorySales/Test/Integration/IsProductSalable/BackorderConditionTest.php:122]: Please do not shorten variable names where possible

File: app/code/Magento/InventoryApi/Test/_files/stocks_with_sales_channel.php [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+/** @var  StockRepositoryInterface */
+$stockRepository = Bootstrap::getObjectManager()->get(StockRepositoryInterface::class);
+
+/** @var State */
+$appState = Bootstrap::getObjectManager()->get(State::class);
+
+/** SalesChannelInterface */
+$salesChannel->setCode('base');
+$salesChannel->setType(SalesChannelInterface::TYPE_WEBSITE);
```
Comment (ishakhsuvarov) [inline app/code/Magento/InventoryApi/Test/_files/stocks_with_sales_channel.php:30]: The fact that this fixture relies on `SalesChannel` hints that it should be in the `InventorySales` module, same as implementation of the fix itself. Please verify if my observations are correct.

**Review by ishakhsuvarov — CHANGES_REQUESTED** [review]

**Pull request closed without merging**